### ISSUE TITLE
Map-style equivalent of `flatmap` - ` FlatMapperMapDataPipe`

### ISSUE THREAD

**Issue by PierreGtch** [pytorch/data#1043]
### 🚀 The feature

The equivalent of ` FlatMapperIterDataPipe` but that would take a `MapDataPipe` as input.
 
This new datapipe would probably need two function parameters as input instead of one: 
- `get_fn` that would return the i-th element of an item in the source datapipe;
- `len_fn` that would return the length of an item in the source datapipe.

See a sketch:
```python
from typing import Callable, Any
from itertools import accumulate

from torch.utils.data import MapDataPipe, functional_datapipe


def default_getitem(d, i):
    return d[i]

@functional_datapipe("flatmap")
class FlatMapperMapDataPipe(MapDataPipe):
    def __init__(self,
                 source_dp: MapDataPipe,
                 get_fn: Callable[[Any, int], Any] = default_getitem,
                 len_fn: Callable[Any, int] = len) -> None:
        super().__init__()
        self.source_dp = source_dp
        self.get_fn = get_fn
        self.len_fn = len_fn

    def _cumsum(self):
        return accumulate(self.len_fn(d) for d in self.source_dp)

    def __len__(self):
        return sum(self.len_fn(d) for d in self.source_dp)

    def _get_source_and_item(self, idx):
        last_size = 0
        for source, size in enumerate(self._cumsum()):
            if size > idx:
                return source, idx - last_size
            last_size = size
        raise IndexError(f'idx {idx} out of range')

    def __getitem__(self, idx):
        if not isinstance(idx, int):
            raise ValueError
        source, i = self._get_source_and_item(idx)
        return self.get_fn(self.source_dp[source], i)
```

### Motivation, pitch

I'm working with mne [raw](https://mne.tools/stable/generated/mne.io.Raw.html) and [epochs](https://mne.tools/stable/generated/mne.Epochs.html) objects to handle heavy EEG recordings.
This library allows you to load only small portions of the long recordings; i.e., you instantiate `epo = mne.Epochs(..., preload=False)` only once, then the different parts of the recording are loaded from disk only when you call `X = epo[i]`.

Using an iter-style flatmap is not the best option because you typically want to randomize the order of your training examples, and a map-style is the natural approach for this kind of object.

### Alternatives

Currently, you have to write a custom datapipe that computes the index of each item according to the length of the different items of the input datapipe.

**Comment by ejguan** [comment]
> Using an iter-style flatmap is not the best option because you typically want to randomize the order of your training examples, and a map-style is the natural approach for this kind of object.

You can always add `shuffle` after iter-style flatmap to achieve random accessing. And, eventually, `mapdatapipe.shuffle()` would convert MapDataPipe to IterDataPipe.

Would the following pipeline work for you?
```py
mapdp = ...  # Your source_dp
dp = mapdp.shuffle()  # Converted to IterDataPipe with shuffled indices
dp = dp.flatmap(map_fn)
dp = dp.shuffle(buffer_size=xxx)
```

**Comment by PierreGtch** [comment]
> > Using an iter-style flatmap is not the best option because you typically want to randomize the order of your training examples, and a map-style is the natural approach for this kind of object.
> 
> You can always add `shuffle` after iter-style flatmap to achieve random accessing. And, eventually, `mapdatapipe.shuffle()` would convert MapDataPipe to IterDataPipe.
> 
> Would the following pipeline work for you?
> 
> ```python
> mapdp = ...  # Your source_dp
> dp = mapdp.shuffle()  # Converted to IterDataPipe with shuffled indices
> dp = dp.flatmap(map_fn)
> dp = dp.shuffle(buffer_size=xxx)  # Shuffle the parts of recordings
> ```

Hi Erjia,

We often have long recordings. With those,  the buffer of the second shuffle would be filled with examples from only a few different recordings, but we would actually be interested in generating batches with examples from diverse recordings.

To create some diversity in the batches in an iter-style, I think we would need a flatmap that would, at every iteration, yield an example from one of the iterables of its pool/buffer (chosen at random). Something like this:

```python
import random
from typing import Callable, Any, TypeVar, Iterator
from itertools import accumulate
from dataclasses import dataclass, field

from torch.utils.data import IterDataPipe, MapDataPipe, functional_datapipe

T_co = TypeVar('T_co', covariant=True)

@functional_datapipe("shuffledflatmap")
@dataclass
class ShuffledFlatMapperIterDataPipe(IterDataPipe):
    datapipe: IterDataPipe[T_co]
    buffer_size: int = 10000
    fn: Callable = lambda x: x
    _buffer: list[T_co] = field(init=False, default_factory=list)
    _rng: random.Random = field(init=False, default_factory=random.Random)

    def __post_init__(self):
        super().__init__()

    def _apply_fn(self, batch):
        return self.fn(batch)

    def __iter__(self) -> Iterator[T_co]:
        for x in self.datapipe:
            while len(self._buffer) == self.buffer_size:
                idx = self._rng.randint(0, len(self._buffer) - 1)
                try:
                    yield next(self._buffer[idx])
                except StopIteration:
                    self._buffer.pop(idx)
            self._buffer.append(iter(self._apply_fn(x)))
        while self._buffer:
            idx = self._rng.randint(0, len(self._buffer) - 1)
            try:
                yield next(self._buffer[idx])
            except StopIteration:
                self._buffer.pop(idx)
```

```python
dp = IterableWrapper([[1, 2, 3, 4], 'abcd', 'ABCD'])
dp = dp.shuffledflatmap(buffer_size=2)
print(list(dp))
# ['a', 'b', 'c', 1, 'd', 'A', 'B', 'C', 2, 'D', 3, 4]
```

**Comment by ejguan** [comment]
@PierreGtch This is actually a good idea. We would love to accept a PR to add such feature. Besides, in order to let `DataLoader2` control the determinism, it would be good if you can add a `set_seed` function.
https://github.com/pytorch/pytorch/blob/4930ae7f82cfee22b2c70778f5c2625379444332/torch/utils/data/datapipes/iter/combinatorics.py#L118-L120

**Comment by PierreGtch** [comment]
@ejguan what is the expected behaviour of `set_shuffle`?

For example here:
```python 
dp = IterableWrapper([[1, 2, 3, 4], 'abcd', 'ABCD'])
dp = dp.shuffled_flatmap()
it1 = iter(dp)


dp.set_shuffle(False)
it2 = iter(dp)

print(list(it2))
# [1, 2, 3, 4, 'a', 'b', 'c', 'd', 'A', 'B', 'C', 'D']

print(list(it1))
# ???
```
Do we want `it1` to return non-shuffled elements because of the call to `set_shuffle` or not?
